icu4c/source/common/cjmlbe.cpp

```diff
@@ -0,0 +1,457 @@
+};
+
+CjMLBreakEngine::CjMLBreakEngine(const UnicodeSet &digitOrOpenPunctuationOrAlphabetSet,
+                                 const UnicodeSet &closePunctuationSet, UErrorCode& status)
+    : fDigitOrOpenPunctuationOrAlphabetSet(digitOrOpenPunctuationOrAlphabetSet),
+      fClosePunctuationSet(closePunctuationSet) {
+    if (U_FAILURE(status)) {
+        return;
+    }
+
+    fUnicodeBlock = new UVector32(status);
+    loadMLmodel(status);
+}
+
+CjMLBreakEngine::~CjMLBreakEngine() {
+    fUnicodeBlock->removeAllElements();
+    delete fUnicodeBlock;
+}
+
+namespace {
+    const char16_t INVALID = u'|';
+}
+
+static UnicodeString* concatChar(const char16_t *str, const UChar32 *arr, UErrorCode &status) {
+    LocalPointer<UnicodeString> result(new UnicodeString(str), status);
+    if (U_FAILURE(status)) { return nullptr; }
+    int32_t idx = 0;
+    while (arr[idx] != -1) {
+        result->append(arr[idx++]);
+    }
+    return result.orphan();
+}
+
+static UnicodeString* cloneString(const UnicodeString &str, UErrorCode &status) {
+    LocalPointer<UnicodeString> result(str.clone(), status);
+    return result.orphan();
+}
+
+static void deleteElement(void *obj) {
+    delete static_cast<Element *>(obj);
+}
+
+static bool isValid(const UnicodeString &s) {
+    return s.length() != 1 || s[0] != INVALID;
+}
+
+
+int32_t CjMLBreakEngine::divideUpDictionaryRange(UText *inText, int32_t rangeStart,
+                                                 int32_t rangeEnd, UVector32 &foundBreaks,
+                                                 const UnicodeString &inString,
+                                                 const LocalPointer<UVector32> &inputMap,
+                                                 UErrorCode &status) const {
+
+    if (U_FAILURE(status)) {
+        return 0;
+    }
+    if (rangeStart >= rangeEnd) {
+        status = U_ILLEGAL_ARGUMENT_ERROR;
+        return 0;
+    }
+
+    UVector32 boundary(inString.countChar32() + 1, status);
+    if (U_FAILURE(status)) {
+        return 0;
+    }
+    int32_t numBreaks = 0;
+    UChar32 ch;
+    UnicodeString index;
+    UVector elementList(&deleteElement, nullptr ,6, status);
+    if (U_FAILURE(status)) {
+        return 0;
+    }
+
+    for (int32_t i = 1; i < inString.length(); i++) {
+        // Remove the first element and append a new element starting from index 4(2+2).
+        if (elementList.size() > 0) {
+            elementList.removeElementAt(0);
+            ch = (i + 2) < inString.length() ? inString.char32At(i + 2) : INVALID;
+            index = (ch != INVALID) ? getUnicodeBlockIndex(ch, status) : UnicodeString(INVALID);
+            elementList.adoptElement(new Element(ch, index), status);
+        }
+        // Initiate the elementList by adding six Element objects.
+        else {
+            initElementList(inString, elementList, status);
+        }
+        if (U_FAILURE(status)) {
+            return 0;
+        }
+
+        evaluateBreakpoint (
```

Comment:
The only place calling evaluateBreakpoint is here, why don't you change the parameter of evaluateBreakpoint to just pass in (elementList, i, numBreaks, boundary, status) and let the evaluateBreakpoint to get the information from elementList ?

Reply:
In the previous multi-thread design, I split these six elements up to avoid race condition for elementList access.
Yes, now that multi-thread is disabled, I should roll back to use elementList only.